packages/flutter/lib/src/widgets/image_filter.dart

```diff
@@ -0,0 +1,60 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ui';
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/rendering.dart';
+
+import 'framework.dart';
+
+/// Applies an [ImageFilter] to its child.
+@immutable
+class ImageFiltered extends SingleChildRenderObjectWidget {
+  /// Creates a widget that applies an [ImageFilter] to its child.
+  ///
+  /// The [imageFilter] must not be null.
+  const ImageFiltered({@required this.imageFilter, Widget child, Key key})
+      : assert(imageFilter != null),
+        super(key: key, child: child);
+
+  /// The image filter to apply to the child of this widget.
+  final ImageFilter imageFilter;
+
+  @override
+  RenderObject createRenderObject(BuildContext context) => _ImageFilterRenderObject(imageFilter);
+
+  @override
+  void updateRenderObject(BuildContext context, _ImageFilterRenderObject renderObject) {
+    renderObject..imageFilter = imageFilter;
+  }
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(DiagnosticsProperty<ImageFilter>('imageFilter', imageFilter));
+  }
+}
+
+class _ImageFilterRenderObject extends RenderProxyBox {
+  _ImageFilterRenderObject(this._imageFilter);
+
+  ImageFilter get imageFilter => _imageFilter;
+  ImageFilter _imageFilter;
+  set imageFilter(ImageFilter value) {
+    assert(value != null);
+    if (value != _imageFilter) {
+      _imageFilter = value;
+      markNeedsPaint();
+    }
+  }
+
+  @override
+  bool get alwaysNeedsCompositing => child != null;
```

Comment:
Can this field change its value dynamically? What happens if on frame 1 `child == null` and on frame 2 `child != null`? Maybe you want [needsCompositing](https://api.flutter.dev/flutter/rendering/RenderObject/needsCompositing.html)?

Reply:
I stole this code from color_filter.dart.  Basically, `sed -s "s/Color/Image/g" < color_filter.dart > image_filter.dart` (but by hand).

Reply:
```
  /// Override this in subclasses to indicate that your paint function always
  /// creates at least one composited layer. For example, videos should return
  /// true if they use hardware decoders.
```

I think the intention is that if I have a child, then the child may require compositing even if I don't, so as long as I have a child I will always need compositing. Otherwise the ROs only notify to be composited when they have a change.

Reply:
You always need compositing if your paint will push a layer. That's all.

Reply:
(and equivalently, you don't "always need compositing" if your paint doesn't need to push a layer)

Compositing is separate from "markNeedsPaint", which is what render objects do when they change. Compositing is about whether ancestor render objects need to use layers when they don't need to. For example, if you have an ancestor clip rect render object, then normally it would just clip on the canvas, but if it has a descendant that needs compositing then instead it'll use a layer. The property is called "always needs compositing" to indicate that you will not chose to _not_ use a layer.